### PR TITLE
Active Learning Tutorial for Tool Tracking Segmentation

### PR DESCRIPTION
**Pull request by finalelement**
Fixes # .

### Description
A few sentences describing the changes proposed in this pull request.

### Checks

- [x] Avoid including large-size files in the PR.
- [x] Clean up long text outputs from code cells in the notebook.
- [ ] For security purposes, please check the contents and remove any sensitive info such as user names and private key.
- [ ] Ensure (1) hyperlinks and markdown anchors are working (2) use relative paths for tutorial repo files (3) put figure and graphs in the `./figure` folder
- [ ] Notebook runs automatically `./runner.sh -t <path to .ipynb file>`

### CODE REVIEW

**Comment by wyli** [comment]
Hi @finalelement, we've updated the formatting requirements for jupyter notebook tutorials, please see https://github.com/Project-MONAI/tutorials/blob/main/CONTRIBUTING.md#create-a-notebook and fixes errors in
- https://github.com/Project-MONAI/tutorials/actions/runs/4823445270/jobs/8591952681?pr=1340
- https://github.com/Project-MONAI/tutorials/actions/runs/4823445275/jobs/8591952686?pr=1340

File: figures/al_robotool_results.xcf
Comment (wyli) [inline figures/al_robotool_results.xcf:1]: is this file needed?

File: auto3dseg/notebooks/auto_runner.ipynb
Comment (wyli) [inline auto3dseg/notebooks/auto_runner.ipynb:1]: I don't think this `auto3dseg/notebooks/auto_runner.ipynb` should be removed

File: active_learning/tool_tracking_al/README.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# Active Learning Simulation Tutorial For Surgical Tool Segmentation
+
+Active learning (AL) is a machine learning framework that can serve as a data selection engine given an unlabeled pool of data. Active learning provides uncertainty estimation capabilities for trained deep learning models, in brief the models are able to assess about which predictions they are under or over confident. Based on the confidence the most relevant data for annotation can be selected. Thereafter annotation the labeled data is added to training pool of data for the model.
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/README.md:3]: Active learning provides uncertainty estimation capabilities for trained deep learning models, in brief the models are able to assess about which predictions they are under or over confident.

This seems two sentences to me.

Comment (mingxin-zheng) [reply active_learning/tool_tracking_al/README.md:3]: annotation the labeled data
->

annotation **of** the labeled data

File: active_learning/tool_tracking_al/README.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+This tutorial provides a toy example of active learning for surgical tool task segmentation. Exploratory notebooks on how to visualize uncertainty/confidence of model as provided as well
+
+### Steps to run the tutorial
+1. Download the data from [here](https://www.synapse.org/#!Synapse:syn22427422)
+2. To just launch the active learning script with 'variance' acquisition function, launch the code with the below command: `python active_learning.py --base_dir /to/be/defined --data_root /to/be/defined --json_path /to/be/defined --seed 120 --active_iters 6 --dropout_ratio 0.2 --mc_number 10 --initial_pool 15 --queries 15 --strategy variance --steps 2000 --val_interval 1 --batch_size 1 --val_batch_size 1 --lr 1e-4`
+3. For executing the active learning acquisition functions of `random`, `variance` and the `full dataset` baseline, execute the `./runner.sh`
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/README.md:10]: I am new to "active learning acquisition functions". Maybe some explanations can be helpful here. Thanks!

File: active_learning/tool_tracking_al/README.md [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+4. Once all script finish running, to analyze results, launch the jupyter notebook `results_uncertainty.ipynb`. The second part particularly will visualize the below shown plot in the results. The first part of the notebook is an exploration of uncertainty spatial maps.
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/README.md:11]: all script**s**

File: active_learning/tool_tracking_al/README.md [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+### 1. Data
+
+[RoboTool dataset](https://www.synapse.org/#!Synapse:syn22427422) [Garcia-Peraza-Herrera et. al 2021] has been used for the purpose of this tutorial. The dataset consists of 514 labeled 2D frames where the surgical tool has been annotated as the foreground mask, the rest of the image has been treated as the background.
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/README.md:15]: [Garcia-Peraza-Herrera et. al 2021]
->
[1]

Comment (mingxin-zheng) [reply active_learning/tool_tracking_al/README.md:15]: The dataset consists of 514 labeled 2D frames where the surgical tool has been annotated as the foreground mask, the rest of the image has been treated as the background.
->
The dataset consists of 514 labeled 2D frames where the surgical tool has been annotated as the foreground mask **and** the rest of the image has been treated as the background.

File: active_learning/tool_tracking_al/README.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+Steps per active iteration: 2000 \
+Batch-size: 1 \
+Learning Rate: 1 \
+Optimizer: Adam \
+Loss: DiceLoss
+
+### 4. Results & Discussion
+
+![al_robotool_results](../../figures/al_robotool_results.png)
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/README.md:40]: The x-axis may be slightly adjusted so that it leaves no blank space on the right side

**Review by wyli — APPROVED** [review]
looks good to me apart from the minor comments  inline

File: active_learning/tool_tracking_al/active_learning.py [hunk 6]
```diff
@@ -0,0 +1,573 @@
...
+
+# Directory & Json & Seed
+parser.add_argument("--base_dir", default="/home/vishwesh/experiments/robo_tool_experiments/variance_sanity", type=str)
+parser.add_argument("--data_root", default="/scratch_2/robo_tool_dataset_2023", type=str)
+parser.add_argument("--json_path", default="/scratch_2/robo_tool_dataset_2023/data_list.json", type=str)
+parser.add_argument("--seed", default=120, type=int)
+
+# Active learning parameters
+parser.add_argument("--active_iters", default=3, type=int)
+parser.add_argument("--dropout_ratio", default=0.2, type=float)
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/active_learning.py:57]: This seems unreferenced in the code

File: active_learning/tool_tracking_al/active_learning.py [hunk 7]
```diff
@@ -0,0 +1,573 @@
...
+parser.add_argument("--initial_pool", default=10, type=int)
+parser.add_argument("--queries", default=10, type=int)
+parser.add_argument("--strategy", default="variance", type=str)  # Options available variance, random
+
+# DL Hyper-parameters
+parser.add_argument("--steps", default=300, type=int)  # Default is 2000, this needs to be tweaked more
+parser.add_argument("--val_interval", default=1, type=int)
+parser.add_argument("--batch_size", default=1, type=int)
+parser.add_argument("--val_batch_size", default=1, type=int)
+parser.add_argument("--lr", default=1e-4, type=float)
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/active_learning.py:68]: This also seems unreferenced in the code

File: active_learning/tool_tracking_al/active_learning.py [hunk 8]
```diff
@@ -0,0 +1,573 @@
...
+        p_json_data = json.load(json_f)
+
+    p_train_data = p_json_data["train"]
+    p_copy_train_data = copy.deepcopy(p_train_data)
+    p_val_data = p_json_data["validation"]
+    p_copy_val_data = copy.deepcopy(p_val_data)
+    p_test_data = p_json_data["test"]
+    p_copy_test_data = copy.deepcopy(p_test_data)
+
+    print("Random Strategy is being used for selection instead of SSL based ranking ...")
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/active_learning.py:114]: Please confirm the random strategy is always applied as I see other options like "variance"

File: active_learning/tool_tracking_al/active_learning.py [hunk 9]
```diff
@@ -0,0 +1,573 @@
...
+    new_json_dict = {}
+    new_json_dict["training"] = p_labeled_data
+    new_json_dict["unlabeled"] = p_copy_unl_d
+    new_json_dict["validation"] = p_copy_val_data
+    new_json_dict["test"] = p_copy_test_data
+
+    init_json_f_path = os.path.join(json_base_dir, "json_init_pool_{}.json".format(args.initial_pool))
+    with open(init_json_f_path, "w") as j_file:
+        json.dump(new_json_dict, j_file)
+    j_file.close()
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/active_learning.py:140]: I think `with open ... as file` can close the file handle.

File: active_learning/tool_tracking_al/runner.sh [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+DATA_ROOT="/to/be/defined"
+JSON_PATH="/to/be/defined"
+LOG_DIR="/to/be/defined"
+
+# Full dataset baseline
+python al_robo_tool.py --base_dir ${LOG_DIR}/all_data_iter5k --data_root ${DATA_ROOT} --json_path ${JSON_PATH} --seed 120 --active_iters 1 --dropout_ratio 0.2 --mc_number 10 --initial_pool 360 --queries 0 --strategy random --steps 5000 --val_interval 1 --batch_size 1 --val_batch_size 1 --lr 1e-4
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/runner.sh:19]: I did not see this script `al_robo_tool.py`. Is this an earlier version of `active_learning.py`. The `dropout_ratio` and `lr` are not referenced in the script by the way.

File: active_learning/tool_tracking_al/active_learning.py [hunk 11]
```diff
@@ -0,0 +1,573 @@
...
+
+        # Define Summary Writer
+        writer = SummaryWriter(log_dir=active_model_dir)
+
+        # Load JSON data
+        if active_iter == 0:
+            print("Opening json file {} for Active iteration {}".format(init_json_f_path, active_iter))
+            with open(init_json_f_path, "rb") as f:
+                json_data = json.load(f)
+            f.close()
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/active_learning.py:235]: Is `f.close()` necessary?

File: active_learning/tool_tracking_al/active_learning.py [hunk 12]
```diff
@@ -0,0 +1,573 @@
...
+        new_json_dict = {}
+        new_json_dict["training"] = copy_train_d
+        new_json_dict["unlabeled"] = copy_unl_d
+        new_json_dict["validation"] = copy_val_d
+        new_json_dict["test"] = copy_test_d
+
+        new_json_file_path = os.path.join(json_base_dir, "json_iter_{}.json".format(active_iter))
+        with open(new_json_file_path, "w") as j_file:
+            json.dump(new_json_dict, j_file)
+        j_file.close()
```
Comment (mingxin-zheng) [inline active_learning/tool_tracking_al/active_learning.py:565]: Same as comment above

File: active_learning/tool_tracking_al/README.md (hunk 1 above)
Comment (finalelement) [reply active_learning/tool_tracking_al/README.md:3]: fixed.

File: active_learning/tool_tracking_al/README.md (hunk 2 above)
Comment (finalelement) [reply active_learning/tool_tracking_al/README.md:10]: I added some explanation

File: active_learning/tool_tracking_al/README.md (hunk 4 above)
Comment (finalelement) [reply active_learning/tool_tracking_al/README.md:15]: fixed

File: active_learning/tool_tracking_al/README.md (hunk 5 above)
Comment (finalelement) [reply active_learning/tool_tracking_al/README.md:40]: will be updating these results again in a separate PR, lets use this for now

File: figures/al_robotool_results.xcf
Comment (finalelement) [reply figures/al_robotool_results.xcf:1]: It's not needed

Comment (finalelement) [reply figures/al_robotool_results.xcf:1]: File has been removed, thanks

File: auto3dseg/notebooks/auto_runner.ipynb
Comment (wyli) [reply auto3dseg/notebooks/auto_runner.ipynb:1]: you can try `git checkout main -- auto3dseg/notebooks/auto_runner.ipynb`

File: active_learning/tool_tracking_al/active_learning.py (hunk 7 above)
Comment (finalelement) [reply active_learning/tool_tracking_al/active_learning.py:68]: This is referenced now.

File: active_learning/tool_tracking_al/active_learning.py (hunk 8 above)
Comment (finalelement) [reply active_learning/tool_tracking_al/active_learning.py:114]: Yes for initial data selection random is always used, irrespective of any given strategy.

**Review by mingxin-zheng — APPROVED** [review]
